### PR TITLE
Implement JSON base wrapper classes

### PR DESCRIPTION
**Pull request by djberg96**
This replaces https://github.com/ManageIQ/azure-armrest/pull/21

With JSON wrapper classes we can get real classes back instead of plain hashes, and add custom methods to them as we see fit.

This PR does not modify the service classes. They still return hashes for now, but you can still use these wrapper classes by calling .to_json first. Basically, I wanted this first step reviewed before altering other code.

Some sample code:

    vmm = Azure::Armrest::VirtualMachineService.new(conf)
    json = vmm.get('miqazure-win2k8', 'miqazure-dan1', true).to_json
    vm = Azure::Armrest::VirtualMachine.new(json)

    p vm.properties.hardware_profile.vm_size # => "Standard_A1"
    p vm.properties.storage_profile.os_disk.os_type # => "Windows"
    p vm.properties.os_profile.computer_name # => "miqazure-win2k8"

### CODE REVIEW

File: lib/azure/armrest/wrapper/base_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+      # A custom Delegator interface method that creates snake_case
+      # versions of the camelCase delegate methods.
+      def __setobj__(obj)
+        obj.methods(false).each{ |m|
+          if m.to_s[-1] != '=' # Must deal with nested ostruct's
+            res = obj.send(m)
+            if res.respond_to?(:each)
+              res.each{ |o| __setobj__(o) if o.is_a?(OpenStruct) }
+            else
```
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: `elsif` is more readable

Comment (djberg96) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Sorry, I don't understand what you mean here.

File: lib/azure/armrest/wrapper/availability_set.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+require_relative 'base_wrapper'
+
+module Azure
+  module Armrest
+    class AvailabilitySet < Azure::ArmRest::BaseWrapper
```
Comment (bzwei) [inline lib/azure/armrest/wrapper/availability_set.rb]: We don't really need to create these many physical files for wrapper classes. Take an example from https://github.com/rest-client/rest-client/blob/master/lib/restclient/exceptions.rb. We can maintain a constant table in the base class and generate wrapper classes by code

Comment (djberg96) [inline lib/azure/armrest/wrapper/availability_set.rb]: The idea here was that we could add custom methods to each class as-needed. It's a lot easier to debug and maintain IMHO.

File: spec/wrappers/event_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+########################################################################
+# event_spec.rb
+#
+# Test suite for the Azure::ArmRest::Event json wrapper class.
+########################################################################
+require 'spec_helper'
+
+describe "Event" do
```
Comment (bzwei) [inline spec/wrappers/event_spec.rb]: If we use code to generate wrapper classes, no need for so many spec tests.

Comment (djberg96) [inline spec/wrappers/event_spec.rb]: As custom methods get added to each class over time, having the separate spec files will be easier to deal with.

File: lib/azure/armrest/wrapper/base_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
+require 'delegate'
+require 'ostruct'
+
+module Azure
+  module ArmRest
+    # Base class for JSON wrapper classes. Each Service class should have
+    # a corresponding class that wraps the JSON it collects, and each of
+    # them should subclass this base class.
+    class BaseWrapper < Delegator
```
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Can we rename `wrapper` to `model` for both the directory and the class name? `Fog` also uses the term `model` although it is not an  `ActiveRecord`

Comment (djberg96) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: You want BaseModel or just Model? I guess I worry about name clashes, but it's not a huge deal.

File: lib/azure/armrest/wrapper/availability_set.rb (hunk 2 above)
Comment (bzwei) [inline lib/azure/armrest/wrapper/availability_set.rb]: Initially we use code to generate wrapper classes. For each individual class if we need to add custom method, then it is the time to physically add a new file.

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 4 above)
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: yes, we can use BaseModel. Should not be a problem, right?

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 1 above)
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: `elsif res.is_a?(OpenStruct)`

File: lib/azure/armrest/wrapper/base_wrapper.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          if m.to_s[-1] != '=' # Must deal with nested ostruct's
+            res = obj.send(m)
+            if res.respond_to?(:each)
+              res.each{ |o| __setobj__(o) if o.is_a?(OpenStruct) }
+            else
+              __setobj__(res) if res.is_a?(OpenStruct)
+            end
+          end
+          snake = m.to_s.gsub(/(.)([A-Z])/,'\1_\2').downcase.to_sym
+          obj.instance_eval("alias #{snake} #{m}")
```
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Thought not hurt, but can we add a check, like `unless snake == m` to avoid most repeated alias?

File: lib/azure/armrest/wrapper/base_wrapper.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        obj.methods(false).each{ |m|
+          if m.to_s[-1] != '=' # Must deal with nested ostruct's
+            res = obj.send(m)
+            if res.respond_to?(:each)
+              res.each{ |o| __setobj__(o) if o.is_a?(OpenStruct) }
+            else
+              __setobj__(res) if res.is_a?(OpenStruct)
+            end
+          end
+          snake = m.to_s.gsub(/(.)([A-Z])/,'\1_\2').downcase.to_sym
```
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Do we have enough evidence the resulting `snake` is a legitimate method name?

File: lib/azure/armrest/resource_provider_service.rb [hunk 7]
```diff
@@ -53,7 +54,8 @@ def list
       def get(namespace)
         url = build_url(namespace)
         response = rest_get(url)
-        JSON.parse(response.body)
+        #JSON.parse(response.body)
+        Azure::Armrest::ResourceProvider.new(response.body)
```
Comment (bzwei) [inline lib/azure/armrest/resource_provider_service.rb]: What's the proper way to convert an array of wrapper objects, for all `list` methods?

Comment (djberg96) [inline lib/azure/armrest/resource_provider_service.rb]: This shouldn't be in there, sorry. I must have been tinkering and left it in by mistake.

For list methods, we'll need to .map the results.

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 5 above)
Comment (djberg96) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Yes.

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 1 above)
Comment (djberg96) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: I'm not sure I follow, since the current code deals with lists.

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 6 above)
Comment (djberg96) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Do you have a counter example?

Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: I don't. We can easily break it with space or special characters in the name, but I doubt Azure API will ever use such strange names.

File: lib/azure/armrest/resource_provider_service.rb (hunk 7 above)
Comment (bzwei) [inline lib/azure/armrest/resource_provider_service.rb]: If the response JSON string from the API contains an array at root level, how do you handle? You don't want to use `JSON.parse` to get an array of hashes, and use `.map` to convert each back to a JSON string, do you?

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 1 above)
Comment (bzwei) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Instead of
```
else
  __setobj__(res) if res.is_a?(OpenStruct)
end
```
we can have
```
elsif res.is_a?(OpenStruct)
  __setobj__(res)
end

File: lib/azure/armrest/wrapper/base_wrapper.rb (hunk 6 above)
Comment (djberg96) [inline lib/azure/armrest/wrapper/base_wrapper.rb]: Ok. If it breaks we can revisit this code, but for now I think it's fine.

File: lib/azure/armrest/resource_provider_service.rb (hunk 7 above)
Comment (djberg96) [inline lib/azure/armrest/resource_provider_service.rb]: The plan is to eventually replace the JSON.parse calls with instances of our new wrapper classes. I thought I would make those changes as a separate PR.
